source/solvers/fluid_dynamics_matrix_free_operators.cc

```diff
@@ -830,6 +986,201 @@ NavierStokesOperatorBase<dim, number>::do_cell_integral_range(
+
+
+
+  // If the boundary condition is not in our list of boundary
+  // condition or the boundary condition that is set in the list is
+  // not a weak function, there is nothing to do and we return
+  if (boundary_id == this->boundary_conditions.id.end() ||
+      this->boundary_conditions.type[boundary_index] !=
+        BoundaryConditions::BoundaryType::function_weak)
+    {
+      const VectorizedArray<number> zero = 0.0;
+
+      for (unsigned int i = 0; i < integrator.dofs_per_cell; ++i)
+        integrator.begin_dof_values()[i] = zero;
+      return;
+    }
+
+
+
+  const auto face = integrator.get_current_cell_index();
+
+  //  std::cout << "Integrating over face " << face << " assemble residual "
+  //            << assemble_residual << " number of quadrature points "
+  //            << integrator.n_q_points << std::endl;
+
+  const double beta = this->boundary_conditions.beta[boundary_index];
+
+
+  const unsigned fe_degree =
+    matrix_free.get_dof_handler().get_fe().tensor_degree();
+
+  VectorizedArray<number> cell_size = 1.0;
+
+  for (unsigned int v = 0;
+       v < matrix_free.n_active_entries_per_face_batch(face);
+       ++v)
+    {
+      const auto [cell_it, _] = matrix_free.get_face_iterator(face, v);
+
+      if (dim == 2)
+        cell_size[v] = std::sqrt(4. * cell_it->measure() / M_PI) / fe_degree;
+      else if (dim == 3)
+        cell_size[v] =
+          std::pow(6 * cell_it->measure() / M_PI, 1. / 3.) / fe_degree;
+    }
+
+  VectorizedArray<number> penalty_parameter =
+    beta / std::pow(cell_size, static_cast<number>(fe_degree + 1));
+
+  integrator.evaluate(EvaluationFlags::EvaluationFlags::values |
+                      EvaluationFlags::EvaluationFlags::gradients);
+
+  for (const auto q : integrator.quadrature_point_indices())
+    {
+      typename FEFaceIntegrator::value_type    value_result    = {};
+      typename FEFaceIntegrator::gradient_type gradient_result = {};
+
+      const auto normal_vector = integrator.get_normal_vector(q);
+
+      const auto value    = integrator.get_value(q);
+      const auto gradient = integrator.get_gradient(q);
+
+
+      for (unsigned int d = 0; d < dim; ++d)
+        { // Assemble (v,beta (u)) for the main penalization
+          value_result[d] += penalty_parameter * value[d];
+
+
+          // Assemble ν(v,∇δu·n)
+          for (unsigned int i = 0; i < dim; ++i)
+            value_result[d] -=
+              kinematic_viscosity * gradient[d][i] * normal_vector[i];
+          //
+          // // // Assemble ν(∇v·n,δu)
+          for (unsigned int i = 0; i < dim; ++i)
+            gradient_result[d][i] -=
+              kinematic_viscosity * value[d] * normal_vector[i];
+        }
+
+
+      // If assemble_residual, we need to also assemble the target value of the
+      // weak dirichlet boundary condition which obtained from the mu_parser
+      // function contained in the boundary_conditions object
+      if constexpr (assemble_residual)
```

Comment:
I would suggest to remove this path. The code below should be able to be merged with the code above if you substract from `value` the value you want to obtain.

Reply:
Great idea, did so.